File: Objective-C/Neocom/Neocom/NCFetchedCollection.m
//
//  NCFetchedCollection.m
//  Neocom
//
//  Created by Artem Shimanski on 23.10.16.
//  Copyright © 2016 Artem Shimanski. All rights reserved.
//

#import "NCFetchedCollection.h"
@import CoreData;

@interface NCFetchedCollection()
@property (strong, nonatomic) NSFetchRequest* fetchRequest;
@end

@implementation NCFetchedCollection

- (id) initWithEntity:(NSString*) entityName predicateFormat:(NSString*) predicateFormat argumentArray:(NSArray*) argumentArray managedObjectContext:(NSManagedObjectContext*) managedObjectContext {
	if (self = [super init]) {
		self.entityName = entityName;
		self.predicateFormat = predicateFormat;
		self.argumentArray = argumentArray ?: @[];
		self.managedObjectContext = managedObjectContext;
		self.fetchRequest = [NSFetchRequest fetchRequestWithEntityName:self.entityName];
		self.fetchRequest.fetchLimit = 1;
	}
	return self;
}

- (id) objectAtIndexedSubscript:(NSInteger) index {
	self.fetchRequest.predicate = [NSPredicate predicateWithFormat:self.predicateFormat argumentArray:[self.argumentArray arrayByAddingObject:@(index)]];
	return [[self.managedObjectContext executeFetchRequest:self.fetchRequest error:nil] lastObject];
}

- (id) objectForKeyedSubscript:(NSString*) key {
	self.fetchRequest.predicate = [NSPredicate predicateWithFormat:self.predicateFormat argumentArray:[self.argumentArray arrayByAddingObject:key ?: [NSNull null]]];
	return [[self.managedObjectContext executeFetchRequest:self.fetchRequest error:nil] lastObject];
}

@end

File: Neocom/Neocom/NCFetchedCollection.swift
//
//  NCFetchedCollection.swift
//  Neocom
//
//  Created by Artem Shimanski on 30.11.16.
//  Copyright © 2016 Artem Shimanski. All rights reserved.
//

import Foundation
import CoreData

public class NCFetchedCollection<Element : NSFetchRequestResult> {
	public let entityName: String
	public let predicateFormat: String
	public let argumentArray:[Any]
	public let managedObjectContext: NSManagedObjectContext
	public let request: NSFetchRequest<Element>
	
	public init(entityName: String, predicateFormat: String, argumentArray:[Any], managedObjectContext: NSManagedObjectContext) {
		self.entityName = entityName
		self.predicateFormat = predicateFormat
		self.argumentArray = argumentArray
		self.managedObjectContext = managedObjectContext
		self.request = NSFetchRequest<Element>(entityName: entityName)
	}
	
	public subscript(index: Int) -> Element? {
		var argumentArray = self.argumentArray
		argumentArray.append(index)
		request.predicate = NSPredicate(format: self.predicateFormat, argumentArray: argumentArray)
		return (try? managedObjectContext.fetch(request))?.last
	}
	
	public subscript(key: String) -> Element? {
		var argumentArray = self.argumentArray
		argumentArray.append(key)
		request.predicate = NSPredicate(format: self.predicateFormat, argumentArray: argumentArray)
		return (try? managedObjectContext.fetch(request))?.last
	}

}
